apps/emqx/test/emqx_persistent_session_ds_fuzzer.erl

```diff
@@ -0,0 +1,741 @@
+%% Proper generators
+%%--------------------------------------------------------------------
+
+qos() ->
+    range(?QOS_0, ?QOS_2).
+
+%% @doc Static part of the client configuration. It is merged with the
+%% randomly generated part.
+%% erlfmt-ignore
+static_client_opts() ->
+    #{
+        port => 1883,
+        proto => v5,
+        clientid => ?clientid,
+        %% These properties are imporant for test logic
+        %%
+        %%   Expiry interval must be large enough to avoid
+        %%   automatic session expiration:
+        properties => #{'Session-Expiry-Interval' => 1000},
+        %%   Clean start is not tested here. It's quite trivial to
+        %%   test session cleanup in a regular test, but
+        %%   accounting for session state resets in the tests is
+        %%   not.
+        clean_start => false,
+        %%   In order for the model to properly account for
+        %%   takeover, clients must not auto-reconnect:
+        reconnect => false,
+        %%   We want to cover as many scenarios where session has
+        %%   un-acked messages as possible:
+        auto_ack => never
+     }.
+
+%% @doc Proper generator for `emqtt:connect' parameters:
+connect_(S) ->
+    ?LET(
+        ReceiveMaximum,
+        range(1, 10),
+        begin
+            DynamicOpts = #{
+                properties => #{'Receive-Maximum' => ReceiveMaximum}
+            },
+            Opts = emqx_utils_maps:deep_merge(static_client_opts(), DynamicOpts),
+            {call, ?MODULE, connect, [S, Opts]}
+        end
+    ).
+
+%% @doc Proper generator that creates a message in one of the topics.
+message(MsgId, #{subs := Subs}) ->
+    %% Create bias towards topics that the session is subscribed to:
+    TopicFreq = [{5, maps:keys(Subs)}, {1, ?topics}],
+    Topics = [{Freq, T} || {Freq, Topics} <- TopicFreq, T <- Topics],
+    ?LET(
+        {Topic, From, QoS},
+        {frequency(Topics), oneof(?publishers), qos()},
+        #message{
+            id = <<>>,
+            qos = QoS,
+            from = From,
+            topic = Topic,
+            %% Note: currently faking time in DS is not trivial. For
+            %% example, when clock of
+            %% `emqx_persistent_session_ds_subs' deviates from
+            %% `emqx_ds_replication_layer''s clock iterators may point
+            %% too far into the future. So we have no choice but to
+            %% use real clock in this test. However, for the sake of
+            %% model determinism we don't assign the timestamp in the
+            %% generator and do it later in the action:
+            timestamp = undefined,
+            %% Message payload is unique:
+            payload = <<Topic/binary, " ", From/binary, " ", (integer_to_binary(MsgId))/binary>>
+        }
+    ).
+
+publish_(S = #{message_seqno := SeqNo}) ->
+    ?LET(
+        BatchSize,
+        range(1, 10),
+        ?LET(
+            Msgs,
+            [message(I, S) || I <- lists:seq(SeqNo, SeqNo + BatchSize)],
```

Comment:
Nit: some clarification would be helpful; for the same thing, we have `MsgId` and `SeqNo` names and `emqx_ds:time()` type.